app/plugins/newgrounds.ts

```diff
@@ -107,23 +165,117 @@ export class NewgroundsPlugin extends BaseSitePlugin {
+        //
+        // (As an alternative for at least Flash videos, there is a script
+        // that does a "new embedController", similar to what happens for
+        // audio files; that script also has the link to the .swf.  This is
+        // not implemented here, but I (E) figured I'd note it.)
+
         const canonicalUrl = this.getCanonicalUrl();
+
+        // This element will be present for both modern and Flash videos
         const videoElt: HTMLVideoElement = querySelector('video');
+        logger.log("newgrounds: video element ", videoElt);
+
+        // This element will only be present for Flash videos
+        const flashVideoElt = querySelector('iframe');
+        logger.log("newgrounds: flash video element ", flashVideoElt);
+
         if (!videoElt || canonicalUrl.indexOf('/portal/view') === -1) {
+            logger.log("newgrounds: no video element, or 'portal/view' not in URL");
             return null;
         }
-        if (getComputedStyle(videoElt, null).visibility === 'hidden') {
-            // This code doesn't work until after the user has started playing the video, as the video element doesn't 
-            // contain a valid video URL until after the user clicks play.
-            // Need to check the visibility of the element first
-            alert("Can't download Newgrounds video unless you've started playing it. Press play and then try again.");
-            return null;
+
+        let fullUrlWithBuster = '';
+
+        // Handle modern videos
+        if(!flashVideoElt) {
+            if (getComputedStyle(videoElt, null).visibility === 'hidden') {
+                // This code doesn't work until after the user has started
+                // playing the video, as the video element doesn't contain a
+                // valid video URL until after the user clicks play.
+                // Need to check the visibility of the element first
+                alert("Can't download Newgrounds video unless you've started playing it. Press play and then try again.");
+                return null;
+            }
+
+            const sourceElt: HTMLSourceElement = videoElt.querySelector('source');
+            fullUrlWithBuster = sourceElt.src;
         }
 
-        const sourceElt: HTMLSourceElement = videoElt.querySelector('source');
-        const fullUrl = sourceElt.src;
+        // Handle Flash videos/games
+        if(flashVideoElt) {
+            // The iframe is available even on the static click-to-play screen,
+            // so try to pull the URL to the .swf out of it.
+            let ruffleLink = flashVideoElt.getAttribute('src');
+            logger.log("newgrounds: ruffle link ", ruffleLink);
+
+            // The parameters to Ruffle are %-encoded
+            let decodedRuffleLink = decodeURI(ruffleLink);
+            logger.log("newgrounds: decoded ruffle link ", decodedRuffleLink);
+            let ruffleParameters = decodedRuffleLink.split('&');
```

Comment:
This mechanism of parsing the URL is brittle and error-prone. Looking at the URL it has, I'd recommend parsing it using the [URL constructor](https://developer.mozilla.org/en-US/docs/Web/API/URL) and then using the [URL.searchParams](https://developer.mozilla.org/en-US/docs/Web/API/URL/searchParams) property to extract the `props` URL param, which you can then parse with [`JSON.parse()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse) to get an object with an `swf` property containing the URL.

Reply:
Are there any Bobby Tables concerns with "trusting" the URL or JSON we pull out of the page and feeding it to the URL or JSON libraries?  Or are the URL and JSON libraries smart enough to stop, throw an exception, or return null if we feed them something naughty?

Reply:
Browsers are *really good* at parsing URLs and JSON correctly and we should let them do it. It's *designed* to be the safe and correct option for these tasks, and both will generally throw if the input is malformed. Browser makers spend a *lot* more time thinking about attacks against these APIs than we ever would with random text parsing code.

(Would you believe in the bad-old-days, we did JSON parsing by `eval()`ing the JSON? *That* was bad security.)